src/librustc_passes/ast_validation.rs

```diff
@@ -35,8 +34,16 @@ impl<'a> AstValidator<'a> {
         &self.session.parse_sess.span_diagnostic
     }
 
+    fn check_lifetime(&self, lifetime: &Lifetime) {
+        if !lifetime.ident.without_first_quote().is_valid() &&
```

Comment:
`lifetime.ident.without_first_quote().is_valid()` -> `token::Ident(ident.node).is_reserved_ident()` and all the new functions (`is_valid`, `is_used_keyword`, `is_static_keyword`, etc) can be removed.
Also `without_first_quote` is used only once so it's probably better inlined right here.

Reply:
Invalid names for lifetimes are all `is_reserved_ident` names except for `keywords::Static` and `keywords::Invalid`.
Invalid names for lifetimes are simply all `is_reserved_ident` plus `'_`.

Reply:
I think I'll still need the `is_static_keyword` and `without_first_quote`: Because the static keyword is reserved as `'static` and the rest isn't, when I check for non-static keywords here and in `check_label` I need to do it removing the first quote and when I check static keyword with it. In the label logic, I don't want to permit the use of `'static` and in the lifetime logic I want to permit it. So in one it should be: `!lifetime.ident.name.is_static_keyword() && !token::Ident(lifetime.ident.without_first_quote()).is_reserved_ident()` and in the other one `lifetime.ident.name.is_static_keyword() || !token::Ident(lifetime.ident.without_first_quote()).is_reserved_ident()`.

Am I right?